File: apps/loomx-api/src/routes/dataSources.ts
import express, { Router } from 'express';
import { metadataProxyService } from '../services/metadataProxy.service';

const router: Router = express.Router();

// GET /api/v1/data-sources - List all data sources with table counts
router.get('/', async (req, res, next) => {
  try {
    // Disable browser caching - LIVE data only
    res.setHeader('Cache-Control', 'no-cache, no-store, must-revalidate');
    res.setHeader('Pragma', 'no-cache');
    res.setHeader('Expires', '0');

    const userEmail = req.headers['x-user-email'] as string || 'unknown';

    const result = await metadataProxyService.query(`
      SELECT
        ds.id,
        ds.name,
        ds.type,
        ds.connection_string,
        ds.database_name,
        ds.region,
        ds.description,
        ds.created_by,
        ds.created_at,
        ds.updated_at,
        ds.is_active,
        CASE WHEN fav.id IS NOT NULL THEN 1 ELSE 0 END AS is_favorite,
        fav.object_id as fav_object_id
      FROM data_sources ds
      LEFT JOIN favorites fav
        ON CAST(ds.id AS NVARCHAR(255)) = fav.object_id
        AND fav.object_type = 'data_source'
        AND fav.user_email = @param0
      ORDER BY
        is_favorite DESC,
        ds.is_active DESC,
        ds.created_at DESC
    `, [userEmail]);

    // Get table counts for each data source using Python proxy
    const { pythonProxyService } = await import('../services/pythonProxy.service');

    const dataSourcesWithCounts = await Promise.all(
      result.rows.map(async (ds: any) => {
        let tableCount = 0;
        try {
          if (ds.database_name) {
            const tables = await pythonProxyService.getTables(ds.database_name);
            tableCount = tables.length;
          }
        } catch (error) {
          console.warn(`[DataSources] Could not get table count for ${ds.name}:`, error);
          // Continue with 0 count on error
        }

        return {
          ...ds,
          table_count: tableCount
        };
      })
    );

    res.json({
      success: true,
      dataSources: dataSourcesWithCounts
    });
  } catch (error) {
    next(error);
  }
});

// GET /api/v1/data-sources/active - List only active data sources with table counts
router.get('/active', async (req, res, next) => {
  try {
    // Disable browser caching - LIVE data only
    res.setHeader('Cache-Control', 'no-cache, no-store, must-revalidate');
    res.setHeader('Pragma', 'no-cache');
    res.setHeader('Expires', '0');

    const userEmail = req.headers['x-user-email'] as string || 'unknown';

    const result = await metadataProxyService.query(`
      SELECT
        ds.id,
        ds.name,
        ds.type,
        ds.connection_string,
        ds.database_name,
        ds.region,
        ds.description,
        ds.created_by,
        ds.created_at,
        ds.updated_at,
        ds.is_active,
        CASE WHEN fav.id IS NOT NULL THEN 1 ELSE 0 END AS is_favorite
      FROM data_sources ds
      LEFT JOIN favorites fav
        ON CAST(ds.id AS NVARCHAR(255)) = fav.object_id
        AND fav.object_type = 'data_source'
        AND fav.user_email = @param0
      WHERE ds.is_active = 1
      ORDER BY
        is_favorite DESC,
        ds.created_at DESC
    `, [userEmail]);

    // Get table counts for each data source using Python proxy
    const { pythonProxyService } = await import('../services/pythonProxy.service');

    const dataSourcesWithCounts = await Promise.all(
      result.rows.map(async (ds: any) => {
        let tableCount = 0;
        try {
          if (ds.database_name) {
            const tables = await pythonProxyService.getTables(ds.database_name);
            tableCount = tables.length;
          }
        } catch (error) {
          console.warn(`[DataSources] Could not get table count for ${ds.name}:`, error);
          // Continue with 0 count on error
        }

        return {
          ...ds,
          table_count: tableCount
        };
      })
    );

    res.json({
      success: true,
      dataSources: dataSourcesWithCounts
    });
  } catch (error) {
    next(error);
  }
});

// GET /api/v1/data-sources/list - List data sources from metadata DB only (no warehouse calls)
// Used by the home page Phase 1 to get data source names/endpoints without table counts.
router.get('/list', async (req, res, next) => {
  try {
    res.setHeader('Cache-Control', 'no-cache, no-store, must-revalidate');

    const userEmail = req.headers['x-user-email'] as string || 'unknown';

    const result = await metadataProxyService.query(`
      SELECT
        ds.id,
        ds.name,
        ds.type,
        ds.connection_string,
        ds.database_name,
        ds.region,
        ds.description,
        ds.is_active,
        CASE WHEN fav.id IS NOT NULL THEN 1 ELSE 0 END AS is_favorite
      FROM data_sources ds
      LEFT JOIN favorites fav
        ON CAST(ds.id AS NVARCHAR(255)) = fav.object_id
        AND fav.object_type = 'data_source'
        AND fav.user_email = @param0
      ORDER BY is_favorite DESC, ds.is_active DESC, ds.created_at DESC
    `, [userEmail]);

    res.json({ success: true, dataSources: result.rows });
  } catch (error) {
    next(error);
  }
});

// GET /api/v1/data-sources/:id - Get a single data source
router.get('/:id', async (req, res, next) => {
  try {
    // Disable browser caching - LIVE data only
    res.setHeader('Cache-Control', 'no-cache, no-store, must-revalidate');
    res.setHeader('Pragma', 'no-cache');
    res.setHeader('Expires', '0');

    const { id } = req.params;

    const result = await metadataProxyService.query(`
      SELECT
        id,
        name,
        type,
        connection_string,
        database_name,
        region,
        description,
        created_by,
        created_at,
        updated_at,
        is_active
      FROM data_sources
      WHERE id = @param0
    `, [parseInt(id)]);

    if (result.rows.length === 0) {
      return res.status(404).json({
        success: false,
        message: 'Data source not found'
      });
    }

    res.json({
      success: true,
      dataSource: result.rows[0]
    });
  } catch (error) {
    next(error);
  }
});

// POST /api/v1/data-sources - Create a new data source
router.post('/', async (req, res, next) => {
  try {
    const { name, type, connection_string, database_name, region, description } = req.body;
    const userEmail = req.headers['x-user-email'] as string;

    // Validate required fields
    if (!name || !type || !connection_string || !region) {
      return res.status(400).json({
        success: false,
        message: 'Missing required fields: name, type, connection_string, region'
      });
    }

    // Validate database_name for Fabric SQL types
    if (type.includes('Fabric SQL') && !database_name) {
      return res.status(400).json({
        success: false,
        message: 'Database name is required for Fabric SQL data sources'
      });
    }

    // Validate region
    if (!['WW', 'EU'].includes(region)) {
      return res.status(400).json({
        success: false,
        message: 'Region must be either "WW" or "EU"'
      });
    }

    const result = await metadataProxyService.query(`
      INSERT INTO data_sources (
        name,
        type,
        connection_string,
        database_name,
        region,
        description,
        created_by,
        is_active
      )
      OUTPUT INSERTED.*
      VALUES (
        @param0,
        @param1,
        @param2,
        @param3,
        @param4,
        @param5,
        @param6,
        1
      )
    `, [name, type, connection_string, database_name || null, region, description || null, userEmail || 'unknown']);

    res.status(201).json({
      success: true,
      dataSource: result.rows[0],
      message: 'Data source created successfully'
    });
  } catch (error: any) {
    // Check for unique constraint violation
    if (error.message && error.message.includes('unique') || error.message.includes('duplicate')) {
      return res.status(409).json({
        success: false,
        message: 'A data source with this name already exists'
      });
    }
    next(error);
  }
});

// PATCH /api/v1/data-sources/:id - Update a data source
router.patch('/:id', async (req, res, next) => {
  try {
    const { id } = req.params;
    const { name, type, connection_string, database_name, region, description, is_active } = req.body;

    // Validate region if provided
    if (region && !['WW', 'EU'].includes(region)) {
      return res.status(400).json({
        success: false,
        message: 'Region must be either "WW" or "EU"'
      });
    }

    // Build dynamic update query
    const updates: string[] = [];
    const params: any[] = [];
    let paramIndex = 0;

    if (name !== undefined) {
      updates.push(`name = @param${paramIndex++}`);
      params.push(name);
    }
    if (type !== undefined) {
      updates.push(`type = @param${paramIndex++}`);
      params.push(type);
    }
    if (connection_string !== undefined) {
      updates.push(`connection_string = @param${paramIndex++}`);
      params.push(connection_string);
    }
    if (database_name !== undefined) {
      updates.push(`database_name = @param${paramIndex++}`);
      params.push(database_name || null);
    }
    if (region !== undefined) {
      updates.push(`region = @param${paramIndex++}`);
      params.push(region);
    }
    if (description !== undefined) {
      updates.push(`description = @param${paramIndex++}`);
      params.push(description || null);
    }
    if (is_active !== undefined) {
      updates.push(`is_active = @param${paramIndex++}`);
      params.push(is_active ? 1 : 0);
    }

    if (updates.length === 0) {
      return res.status(400).json({
        success: false,
        message: 'No fields to update'
      });
    }

    updates.push('updated_at = GETDATE()');
    params.push(parseInt(id)); // Add id as last parameter

    const result = await metadataProxyService.query(`
      UPDATE data_sources
      SET ${updates.join(', ')}
      OUTPUT INSERTED.*
      WHERE id = @param${paramIndex}
    `, params);

    if (result.rows.length === 0) {
      return res.status(404).json({
        success: false,
        message: 'Data source not found'
      });
    }

    res.json({
      success: true,
      dataSource: result.rows[0],
      message: 'Data source updated successfully'
    });
  } catch (error: any) {
    // Check for unique constraint violation
    if (error.message && (error.message.includes('unique') || error.message.includes('duplicate'))) {
      return res.status(409).json({
        success: false,
        message: 'A data source with this name already exists'
      });
    }
    next(error);
  }
});

// DELETE /api/v1/data-sources/:id - Delete a data source
router.delete('/:id', async (req, res, next) => {
  try {
    const { id } = req.params;

    const result = await metadataProxyService.query(`
      DELETE FROM data_sources
      OUTPUT DELETED.*
      WHERE id = @param0
    `, [parseInt(id)]);

    if (result.rows.length === 0) {
      return res.status(404).json({
        success: false,
        message: 'Data source not found'
      });
    }

    res.json({
      success: true,
      message: 'Data source deleted successfully',
      dataSource: result.rows[0]
    });
  } catch (error) {
    next(error);
  }
});

// GET /api/v1/data-sources/:id/table-count - Get table count for a data source
// SIMPLIFIED: Just return null for now to avoid connection overhead
router.get('/:id/table-count', async (req, res, next) => {
  try {
    // Disable browser caching - LIVE data only
    res.setHeader('Cache-Control', 'no-cache, no-store, must-revalidate');
    res.setHeader('Pragma', 'no-cache');
    res.setHeader('Expires', '0');

    // Return null immediately - table counting can be added later when needed
    res.json({
      success: true,
      tableCount: null,
      message: 'Table counting disabled for performance'
    });
  } catch (error) {
    next(error);
  }
});

// POST /api/v1/data-sources/:id/test - Test connection to a data source
router.post('/:id/test', async (req, res, next) => {
  try {
    const { id } = req.params;

    // Get the data source
    const dataSourceResult = await metadataProxyService.query(`
      SELECT connection_string, database_name, type
      FROM data_sources
      WHERE id = @param0
    `, [parseInt(id)]);

    if (dataSourceResult.rows.length === 0) {
      return res.status(404).json({
        success: false,
        message: 'Data source not found'
      });
    }

    const { connection_string, database_name, type } = dataSourceResult.rows[0];

    // TODO: Implement actual connection test
    // For now, just return a success message
    res.json({
      success: true,
      message: 'Connection test not yet implemented',
      connectionString: connection_string,
      database: database_name,
      type
    });
  } catch (error) {
    next(error);
  }
});

// POST /api/v1/data-sources/:id/favorite - Set as favorite data source
router.post('/:id/favorite', async (req, res, next) => {
  try {
    const { id } = req.params;
    const userEmail = req.headers['x-user-email'] as string;

    if (!userEmail) {
      return res.status(401).json({
        success: false,
        message: 'User email required'
      });
    }

    // Check if data source exists
    const dsCheck = await metadataProxyService.query(`
      SELECT id FROM data_sources WHERE id = @param0
    `, [parseInt(id)]);

    if (dsCheck.rows.length === 0) {
      return res.status(404).json({
        success: false,
        message: 'Data source not found'
      });
    }

    // Get the data source name for the favorites record
    const dsResult = await metadataProxyService.query(`
      SELECT name FROM data_sources WHERE id = @param0
    `, [parseInt(id)]);
    const dataSourceName = dsResult.rows[0]?.name || 'Unknown';

    // Delete any existing data source favorites for this user (only one favorite data source allowed)
    await metadataProxyService.query(`
      DELETE FROM favorites
      WHERE user_email = @param0
        AND object_type = 'data_source'
    `, [userEmail]);

    // Set new favorite (object_id is NVARCHAR, so pass as string)
    await metadataProxyService.query(`
      INSERT INTO favorites (user_email, object_id, object_type, object_name, created_at)
      VALUES (@param0, @param1, 'data_source', @param2, GETDATE())
    `, [userEmail, id.toString(), dataSourceName]);

    res.json({
      success: true,
      message: 'Data source set as favorite'
    });
  } catch (error) {
    console.error('[Favorite] Error setting favorite:', error);
    next(error);
  }
});

// DELETE /api/v1/data-sources/:id/favorite - Remove favorite
router.delete('/:id/favorite', async (req, res, next) => {
  try {
    const { id } = req.params;
    const userEmail = req.headers['x-user-email'] as string;

    if (!userEmail) {
      return res.status(401).json({
        success: false,
        message: 'User email required'
      });
    }

    await metadataProxyService.query(`
      DELETE FROM favorites
      WHERE user_email = @param0
        AND object_id = @param1
        AND object_type = 'data_source'
    `, [userEmail, id.toString()]);

    res.json({
      success: true,
      message: 'Favorite removed'
    });
  } catch (error) {
    next(error);
  }
});

// GET /api/v1/data-sources/favorite/current - Get current user's favorite data source
router.get('/favorite/current', async (req, res, next) => {
  try {
    const userEmail = req.headers['x-user-email'] as string;

    if (!userEmail) {
      return res.status(401).json({
        success: false,
        message: 'User email required'
      });
    }

    const result = await metadataProxyService.query(`
      SELECT
        ds.id,
        ds.name,
        ds.type,
        ds.connection_string,
        ds.database_name,
        ds.region,
        ds.description,
        ds.is_active
      FROM favorites fav
      INNER JOIN data_sources ds ON fav.object_id = CAST(ds.id AS NVARCHAR(255))
      WHERE fav.user_email = @param0
        AND fav.object_type = 'data_source'
    `, [userEmail]);

    if (result.rows.length === 0) {
      return res.json({
        success: true,
        dataSource: null
      });
    }

    res.json({
      success: true,
      dataSource: result.rows[0]
    });
  } catch (error) {
    next(error);
  }
});

export default router;

File: apps/kaveon-api/routers/data_sources.py
"""Data sources router — /api/v1/data-sources."""

from fastapi import APIRouter, Request, Response, HTTPException, Depends
from middleware.auth import require_auth
from middleware.permissions import require_min_role
import database.metadata as db
import database.pool as pool

router = APIRouter()
NO_CACHE = {
    "Cache-Control": "no-cache, no-store, must-revalidate",
    "Pragma": "no-cache",
    "Expires": "0",
}

# connection_string is never returned to clients — it may contain credentials
_PUBLIC_FIELDS = "ds.id, ds.name, ds.type, ds.database_name, ds.region, ds.description, ds.created_by, ds.created_at, ds.updated_at, ds.is_active"

_LIST_SELECT = f"""
    SELECT {_PUBLIC_FIELDS},
           CASE WHEN fav.id IS NOT NULL THEN 1 ELSE 0 END AS is_favorite
    FROM data_sources ds
    LEFT JOIN favorites fav
      ON CAST(ds.id AS NVARCHAR(255)) = fav.object_id
      AND fav.object_type = 'data_source' AND fav.user_email = @param0
    ORDER BY is_favorite DESC, ds.is_active DESC, ds.created_at DESC
"""


def _add_table_counts(data_sources: list) -> list:
    results = []
    for ds in data_sources:
        count = 0
        try:
            if ds.get("database_name"):
                tables = pool.get_tables(ds["database_name"])
                count = len(tables)
        except Exception:
            pass
        results.append({**ds, "table_count": count})
    return results


@router.get("/data-sources")
def list_data_sources(request: Request, response: Response, user: str = Depends(require_auth)):
    response.headers.update(NO_CACHE)
    result = db.query(_LIST_SELECT, [user])
    return {"success": True, "dataSources": _add_table_counts(result["rows"])}


@router.get("/data-sources/active")
def list_active_data_sources(request: Request, response: Response, user: str = Depends(require_auth)):
    response.headers.update(NO_CACHE)
    result = db.query(f"""
        SELECT {_PUBLIC_FIELDS},
               CASE WHEN fav.id IS NOT NULL THEN 1 ELSE 0 END AS is_favorite
        FROM data_sources ds
        LEFT JOIN favorites fav
          ON CAST(ds.id AS NVARCHAR(255)) = fav.object_id
          AND fav.object_type = 'data_source' AND fav.user_email = @param0
        WHERE ds.is_active = 1
        ORDER BY is_favorite DESC, ds.created_at DESC
    """, [user])
    return {"success": True, "dataSources": _add_table_counts(result["rows"])}


@router.get("/data-sources/list")
def list_data_sources_metadata_only(request: Request, response: Response, user: str = Depends(require_auth)):
    response.headers["Cache-Control"] = "no-cache, no-store, must-revalidate"
    result = db.query(f"""
        SELECT {_PUBLIC_FIELDS},
               CASE WHEN fav.id IS NOT NULL THEN 1 ELSE 0 END AS is_favorite
        FROM data_sources ds
        LEFT JOIN favorites fav
          ON CAST(ds.id AS NVARCHAR(255)) = fav.object_id
          AND fav.object_type = 'data_source' AND fav.user_email = @param0
        ORDER BY is_favorite DESC, ds.is_active DESC, ds.created_at DESC
    """, [user])
    return {"success": True, "dataSources": result["rows"]}


@router.get("/data-sources/favorite/current")
def get_favorite_data_source(user: str = Depends(require_auth)):
    result = db.query(f"""
        SELECT {_PUBLIC_FIELDS}
        FROM favorites fav
        INNER JOIN data_sources ds ON fav.object_id = CAST(ds.id AS NVARCHAR(255))
        WHERE fav.user_email = @param0 AND fav.object_type = 'data_source'
    """, [user])
    return {"success": True, "dataSource": result["rows"][0] if result["rows"] else None}


@router.get("/data-sources/{ds_id}/table-count")
def get_table_count(ds_id: str, response: Response, user: str = Depends(require_auth)):
    response.headers.update(NO_CACHE)
    return {"success": True, "tableCount": None, "message": "Table counting disabled for performance"}


@router.get("/data-sources/{ds_id}")
def get_data_source(ds_id: str, response: Response, user: str = Depends(require_auth)):
    response.headers.update(NO_CACHE)
    result = db.query(
        f"SELECT {_PUBLIC_FIELDS} FROM data_sources ds WHERE ds.id = @param0",
        [int(ds_id)]
    )
    if not result["rows"]:
        raise HTTPException(status_code=404, detail="Data source not found")
    return {"success": True, "dataSource": result["rows"][0]}


@router.post("/data-sources", status_code=201)
def create_data_source(data: dict, ctx=Depends(require_min_role("Admin"))):
    user = ctx.email
    name = data.get("name")
    ds_type = data.get("type")
    connection_string = data.get("connection_string")
    region = data.get("region")
    if not name or not ds_type or not connection_string or not region:
        raise HTTPException(status_code=400, detail="Missing required fields: name, type, connection_string, region")
    if ds_type != "StarRocks" and not data.get("database_name"):
        raise HTTPException(status_code=400, detail=f"Database name is required for {ds_type} data sources")
    if region not in ("WW", "EU"):
        raise HTTPException(status_code=400, detail='Region must be either "WW" or "EU"')

    try:
        # Use OUTPUT INSERTED but only return public fields (not connection_string)
        db.execute("""
            INSERT INTO data_sources (name, type, connection_string, database_name,
                                      region, description, created_by, is_active)
            VALUES (@param0, @param1, @param2, @param3, @param4, @param5, @param6, 1)
        """, [name, ds_type, connection_string, data.get("database_name"), region,
              data.get("description"), user])
        inserted = db.query_one(
            f"SELECT TOP 1 {_PUBLIC_FIELDS} FROM data_sources ds WHERE ds.name = @param0 AND ds.created_by = @param1 ORDER BY ds.id DESC",
            [name, user]
        )
        return {"success": True, "dataSource": inserted, "message": "Data source created successfully"}
    except Exception as e:
        msg = str(e).lower()
        if "unique" in msg or "duplicate" in msg:
            raise HTTPException(status_code=409, detail="A data source with this name already exists")
        raise HTTPException(status_code=500, detail="Failed to create data source")


@router.patch("/data-sources/{ds_id}")
def update_data_source(ds_id: str, data: dict, ctx=Depends(require_min_role("Admin"))):
    user = ctx.email  # noqa: F841
    region = data.get("region")
    if region and region not in ("WW", "EU"):
        raise HTTPException(status_code=400, detail='Region must be either "WW" or "EU"')

    updates, params, i = [], [], 0
    for field, col in [("name", "name"), ("type", "type"),
                       ("connection_string", "connection_string"),
                       ("database_name", "database_name"), ("region", "region"),
                       ("description", "description")]:
        if field in data:
            updates.append(f"{col} = @param{i}"); params.append(data[field] or None); i += 1
    if "is_active" in data:
        updates.append(f"is_active = @param{i}"); params.append(1 if data["is_active"] else 0); i += 1

    if not updates:
        raise HTTPException(status_code=400, detail="No fields to update")

    updates.append("updated_at = GETDATE()")
    params.append(int(ds_id))

    try:
        count = db.execute(
            f"UPDATE data_sources SET {', '.join(updates)} WHERE id = @param{i}",
            params
        )
        if not count:
            raise HTTPException(status_code=404, detail="Data source not found")
        updated = db.query_one(f"SELECT {_PUBLIC_FIELDS} FROM data_sources ds WHERE ds.id = @param0", [int(ds_id)])
        return {"success": True, "dataSource": updated, "message": "Data source updated successfully"}
    except HTTPException:
        raise
    except Exception as e:
        msg = str(e).lower()
        if "unique" in msg or "duplicate" in msg:
            raise HTTPException(status_code=409, detail="A data source with this name already exists")
        raise HTTPException(status_code=500, detail="Failed to update data source")


@router.delete("/data-sources/{ds_id}")
def delete_data_source(ds_id: str, ctx=Depends(require_min_role("Admin"))):
    count = db.execute("DELETE FROM data_sources WHERE id = @param0", [int(ds_id)])
    if not count:
        raise HTTPException(status_code=404, detail="Data source not found")
    return {"success": True, "message": "Data source deleted successfully"}


@router.post("/data-sources/{ds_id}/test")
def test_data_source(ds_id: str, user: str = Depends(require_auth)):
    row = db.query_one("SELECT database_name, type FROM data_sources WHERE id = @param0", [int(ds_id)])
    if not row:
        raise HTTPException(status_code=404, detail="Data source not found")
    return {"success": True, "message": "Connection test not yet implemented",
            "database": row.get("database_name"), "type": row.get("type")}


@router.post("/data-sources/{ds_id}/favorite")
def set_ds_favorite(ds_id: str, user: str = Depends(require_auth)):
    ds = db.query_one("SELECT id, name FROM data_sources WHERE id = @param0", [int(ds_id)])
    if not ds:
        raise HTTPException(status_code=404, detail="Data source not found")
    db.execute("DELETE FROM favorites WHERE user_email = @param0 AND object_type = 'data_source'", [user])
    db.execute(
        "INSERT INTO favorites (user_email, object_id, object_type, object_name, created_at) "
        "VALUES (@param0, @param1, 'data_source', @param2, GETDATE())",
        [user, str(ds_id), ds.get("name", "Unknown")]
    )
    return {"success": True, "message": "Data source set as favorite"}


@router.delete("/data-sources/{ds_id}/favorite")
def remove_ds_favorite(ds_id: str, user: str = Depends(require_auth)):
    db.execute(
        "DELETE FROM favorites WHERE user_email = @param0 AND object_id = @param1 AND object_type = 'data_source'",
        [user, str(ds_id)]
    )
    return {"success": True, "message": "Favorite removed"}
